### PR TITLE
[CIT-334] Add How to build doc

### PR DESCRIPTION
**Pull request by makr11st**
Signed-off-by: Lukasz Woznicki <lukasz.woznicki@softwareag.com>

## Proposed changes

Add How to Build document.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: BUILDING.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+thin-edge.io operates the `MSRV` (Minimum Supported Rust Version) and uses stable toolchain.
+
+Current MSRV is `1.46.0`.
+
+## Compiling
+
+To build thin-edge.io we are using `cargo`.
+
+As we are using a cargo workspace for all our crates all binaries are put in `./target/` directory with target's name eg: `./target/debug` or `./target/armv7-unknown-linux-gnueabihf` dependent on the type of build.
+
```
Comment (PradeepKiruvale) [inline BUILDING.md:39]: is not it ./target/debug or ./target/release? then armv7*

File: BUILDING.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+To compile in dev mode we use following command:
+
+```shell
+cargo build --release
+```
+
+Build artifacts can be found in `./target/release` and will include executables:
+
+```shell
+$ ls ./target/debug/ted*
```
Comment (PradeepKiruvale) [inline BUILDING.md:71]: should be ./target/release?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -9,10 +9,12 @@ Of course, contributing code is more than welcome! If you're planning to submit
 
 If you are interested in contributing documentation, please note the following:
 
-
 - Doc issues are labeled with the `doc` label.
 - The thin-edge.io docs content is in the `docs/src/` directory.
 
+[How to build from source.](./BUILDING.md)
+
+<br/>
```
Comment (PradeepKiruvale) [inline CONTRIBUTING.md:17]: extra line can be removed

File: BUILDING.md (hunk 1 above)
Comment (makr11st) [reply BUILDING.md:39]: No, for cross builds it goes to `./target/arch_name/profile_name`.
Maybe I should just add separate section for cross compilation to not introduce confusion.

**Review by PradeepKiruvale — APPROVED** [review]
LGTM

File: README.md [hunk 4]
```diff
@@ -39,25 +35,22 @@
 <br/>
 <br/>
 
-
 <!-- ABOUT THE PROJECT -->
 # About The Project
 
-thin-edge.io is the first open-source and cloud-agnostic edge framework designed for resource-constrained IoT edge devices.  
+thin-edge.io is the first open-source and cloud-agnostic edge framework designed for resource-constrained IoT edge devices.
```
Comment (rina23q) [inline README.md:41]: There is a NBSP between `for` and  `resource-constrained`.

File: README.md [hunk 5]
```diff
@@ -39,25 +35,22 @@
...
 
-We have started the project to solve the challenge of IoT device enablement in industrial IoT or more precisely to make the integration of IoT devices with cloud and IoT platforms easy and reliable.  
+We have started the project to solve the challenge of IoT device enablement in industrial IoT or more precisely to make the integration of IoT devices with cloud and IoT platforms easy and reliable.
```
Comment (rina23q) [inline README.md:43]: Two NBSPs between
- `make` and `the`
- `and` and  `reliable`
Some editor make the NBSPs visible. From GH, we can see the NBSP looks narrower gap than the normal spaces.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -9,10 +9,12 @@ Of course, contributing code is more than welcome! If you're planning to submit
 
 If you are interested in contributing documentation, please note the following:
 
-
 - Doc issues are labeled with the `doc` label.
 - The thin-edge.io docs content is in the `docs/src/` directory.
 
+[How to build from source.](./BUILDING.md)
```
Comment (rina23q) [inline CONTRIBUTING.md:15]: I think we want to have discussion where to give the link to the doc how to build. For me, we can put it in README, where everyone can see it for the first.

File: BUILDING.md (hunk 1 above)
Comment (rina23q) [reply BUILDING.md:48]: Using `<>` for architecture may help people to understand slightly better? 
`./target/<architecture>/debug`, `./target/<architecture>/release`

File: BUILDING.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+> For other shells, you'll want to find the appropriate place to set a configuration at start time,
+> eg. zsh uses ~/.zshrc. Check your shell's documentation to find what file it uses.
+
+thin-edge.io operates the `MSRV` (Minimum Supported Rust Version) and uses stable toolchain.
+
+Current MSRV is `1.46.0`.
+
+### Cross compilation toolchain
+
+thin-edge.io can be compiled for target architecture on non-target device, this is called cross compilation. Currently we support Raspberry Pi 3 building for `armv7` architecture with Rust's cross compilation toolchain called `cargo cross`.
```
Comment (rina23q) [inline BUILDING.md:36]: Raspberry Pi 3 => Raspberry Pi 3B. Actually, there is Raspberry Pi 3A, which I've never seen someone is using :D

File: BUILDING.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+To compile dev profile (with debug symbols) we use following command:
+
+```shell
+cargo build
+```
+
+Build artifacts can be found in `./target/debug` and will include executables:
+
+```shell
+$ ls ./target/debug/ted*
```
Comment (rina23q) [inline BUILDING.md:60]: Looks lazy lol
```suggestion
$ ls ./target/debug/tedge*
```

File: BUILDING.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+To create binaries which can run on different platform than one you are currently on you can use `cargo cross`:
+
+```shell
+cross build --target armv7-unknown-linux-gnueabihf
+```
+
+Build artifacts can be found in `./target/armv7-unknown-linux-gnueabihf/debug` and will include executables:
+
+```shell
+$ ls ./target/armv7-unknown-linux-gnueabihf/debug/ted*
```
Comment (rina23q) [inline BUILDING.md:101]: Same as above.
```suggestion
$ ls ./target/armv7-unknown-linux-gnueabihf/debug/tedge*
```

**Review by rina23q — CHANGES_REQUESTED** [review]
Some change. Especially, I want to talk about moving the link to this guide to README from CONTRIBUTION.

File: README.md (hunk 4 above)
Comment (makr11st) [reply README.md:41]: nbsp is not markdown standard, it is custom addition for markdown interpreters which support HTML.

File: CONTRIBUTING.md (hunk 6 above)
Comment (makr11st) [reply CONTRIBUTING.md:15]: It wasn't specified anywhere where the link should go, I just chose a place I felt was good enough. As our README has nothing like 'quick start' or 'how to' it just didn't feel right to place it there as our main intention is for users to use as packages from releases and not build it from source.
Maybe Contributing section of readme would be good.

File: README.md (hunk 4 above)
Comment (rina23q) [reply README.md:42]: Shouldn't we delete NBSP?

File: CONTRIBUTING.md (hunk 6 above)
Comment (rina23q) [reply CONTRIBUTING.md:15]: As Sebastian said, putting the links in both in README and CONTRIBUTING make sense. I agree that contributing section of readme is a good place.

**Review by rina23q — APPROVED** [review]
Thanks Lukasz!

**Review by itsyitsy — APPROVED** [review]

**Pull request merged**